### PR TITLE
make  rules engine code public

### PR DESCRIPTION
**Pull request by shalehaha**
## Description

1. Make LaunchRulesEngine public, with methods:
    * `public func replaceRules(with rules: [LaunchRule]) `
    * `public func process(event: Event) -> Event`
    * `func replaceRules(from urlString: String)`
    * `func replaceRulesWithCache(from urlString: String)`
2. Make JSONRulesParser public
3. Make LaunchRule and RuleConsequnce class public.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: AEPCore/Sources/rules/LaunchRulesEngine.swift [hunk 1]
```diff
@@ -12,11 +12,11 @@
 
 import AEPServices
 import Foundation
-@_implementationOnly import AEPRulesEngine
+import AEPRulesEngine
 
 /// A rules engine for Launch rules
-class LaunchRulesEngine {
-    private let LOG_TAG = "\(RulesConstants.LOG_MODULE_PREFIX) - LaunchRulesEngine"
+public class LaunchRulesEngine {
```
Comment (shalehaha) [inline AEPCore/Sources/rules/LaunchRulesEngine.swift:18]: need to find a better name for `LaunchRulesEngine`

File: AEPCore/Sources/rules/LaunchRuleConsequence.swift [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ governing permissions and limitations under the License.
+ */
+
+import Foundation
+import AEPRulesEngine
+
+public struct RuleConsequence {
+    let id: String
+    let type: String
+    var detailDict: [String: Any?]
```
Comment (sbenedicadb) [inline AEPCore/Sources/rules/LaunchRuleConsequence.swift:19]: can we change this var name to `details`?

File: AEPCore/Sources/rules/LaunchRuleConsequence.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/*
+ Copyright 2020 Adobe. All rights reserved.
```
Comment (sbenedicadb) [inline AEPCore/Sources/rules/LaunchRuleConsequence.swift:2]: 2021

File: AEPCore/Sources/rules/LaunchRulesEngine.swift [hunk 4]
```diff
@@ -30,31 +30,26 @@ class LaunchRulesEngine {
...
+    private let rulesQueue: DispatchQueue
+    private var waitingEvents: [Event]?
     private let dataStore: NamedCollectionDataStore
 
     let evaluator: ConditionEvaluator
     let rulesEngine: RulesEngine<LaunchRule>
-    let rulesDownloader: RulesDownloader
 
-    init(name: String, extensionRuntime: ExtensionRuntime) {
+    public init(name: String, extensionRuntime: ExtensionRuntime) {
```
Comment (sbenedicadb) [inline AEPCore/Sources/rules/LaunchRulesEngine.swift:40]: would be good to get some in-code documentation here, especially for public apis

File: AEPCore/Sources/rules/LaunchRulesEngine+Downloader.swift [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            return
+        }
+        let rulesDownloader = RulesDownloader(fileUnzipper: FileUnzipper())
+        rulesDownloader.loadRulesFromUrl(rulesUrl: url) { data in
+            guard let data = data else {
+                return
+            }
+
+            guard let rules = JSONRulesParser.parse(data) else {
+                return
```
Comment (yangyansong-adbe) [inline AEPCore/Sources/rules/LaunchRulesEngine+Downloader.swift:33]: We'd better combine two `guard`s and print warning log here. 
```Swift
guard let data = data, let rules = JSONRulesParser.parse(data)  else {
                 // print warning log here.
                 return
             }
```

File: AEPCore/Sources/rules/LaunchRulesEngine+Downloader.swift [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            Log.warning(label: RulesConstants.LOG_MODULE_PREFIX, "Invalid rules url: \(urlString)")
+            return
+        }
+        let rulesDownloader = RulesDownloader(fileUnzipper: FileUnzipper())
+        guard let data = rulesDownloader.loadRulesFromCache(rulesUrl: url) else {
+            return
+        }
+
+        guard let rules = JSONRulesParser.parse(data) else {
+            return
```
Comment (yangyansong-adbe) [inline AEPCore/Sources/rules/LaunchRulesEngine+Downloader.swift:53]: need to combine `guard`s and print log.

**Review by yangyansong-adbe — APPROVED** [review]

**Pull request merged**